hoodie-common/src/main/java/com/uber/hoodie/common/util/collection/ExternalSpillableMap.java

```diff
@@ -142,51 +136,60 @@ public boolean containsValue(Object value) {
 
   @Override
   public R get(Object key) {
-    if(inMemoryMap.containsKey(key)) {
+    if (inMemoryMap.containsKey(key)) {
       return inMemoryMap.get(key);
-    } else if(diskBasedMap.containsKey(key)) {
+    } else if (diskBasedMap.containsKey(key)) {
       return diskBasedMap.get(key);
     }
     return null;
   }
 
   @Override
   public R put(T key, R value) {
-    try {
-      if (this.currentInMemoryMapSize < maxInMemorySizeInBytes || inMemoryMap.containsKey(key)) {
-        // Naive approach for now
-        if (estimatedPayloadSize == 0) {
-          this.estimatedPayloadSize = SpillableMapUtils.computePayloadSize(value, schema);
-          log.info("Estimated Payload size => " + estimatedPayloadSize);
-        }
-        if(!inMemoryMap.containsKey(key)) {
-          currentInMemoryMapSize += this.estimatedPayloadSize;
-        }
-        inMemoryMap.put(key, value);
-      } else {
-        diskBasedMap.put(key, value);
+    if (this.currentInMemoryMapSize < maxInMemorySizeInBytes || inMemoryMap.containsKey(key)) {
+      if (shouldEstimatePayloadSize && estimatedPayloadSize == 0) {
+        // At first, use the sizeEstimate of a record being inserted into the spillable map.
+        // Note, the converter may over estimate the size of a record in the JVM
+        this.estimatedPayloadSize =
+            keyConverter.sizeEstimate(key) + valueConverter.sizeEstimate(value);
+        log.info("Estimated Payload size => " + estimatedPayloadSize);
+      }
+      else if(shouldEstimatePayloadSize &&
+          inMemoryMap.size() % NUMBER_OF_RECORDS_TO_ESTIMATE_PAYLOAD_SIZE == 0) {
+        // Re-estimate the size of a record by calculating the size of the entire map containing
+        // N entries and then dividing by the number of entries present (N). This helps to get a
+        // correct estimation of the size of each record in the JVM.
+        long totalMapSize = ObjectSizeCalculator.getObjectSize(inMemoryMap);
+        this.currentInMemoryMapSize = totalMapSize;
+        this.estimatedPayloadSize = totalMapSize/inMemoryMap.size();
+        shouldEstimatePayloadSize = false;
+        log.info("New Estimated Payload size => " + this.estimatedPayloadSize);
       }
-      return value;
-    } catch(IOException io) {
-      throw new HoodieIOException("Unable to estimate size of payload", io);
+      if (!inMemoryMap.containsKey(key)) {
```

Comment:
I think we are assuming records are of the same size here and updates need not adjust for size..  worthy to leave a TODO here to revisit..

Reply:
Yes, we are assuming that, done.